cpp/src/arrow/filesystem/filesystem.h

```diff
@@ -0,0 +1,145 @@
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <chrono>
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "arrow/status.h"
+
+namespace arrow {
+
+namespace io {
+
+class InputStream;
+class OutputStream;
+
+}  // namespace io
+
+namespace fs {
+
+// A system clock time point expressed as a 64-bit (or more) number of
+// secondes since the epoch.
+using TimePoint = std::chrono::sys_time<std::chrono::nanoseconds>;
+
+enum class FileType {
+  // Target does not exist
+  NonExistent,
+  // Target exists but its type is unknown (could be a special file such
+  // as a Unix socket or character device, or Windows NUL / CON / ...)
+  Unknown,
+  // Target is a regular file
+  File,
+  // Target is a directory
+  Directory
+};
+
+std::string ToString(FileType);
+
+struct FileStats {
+  FileType type = FileType::Unknown;
+  // The file path in the filesystem.
+  std::string path;
+  // The size in bytes, if available.  Only regular files are guaranteed
+  // to have a size.
+  int64_t size = -1;
+  // The time of last modification, if available.
+  TimePoint mtime = TimePoint(-1);
+
+  FileStats() = default;
+
+  // The file base name (component after the last '/').
+  std::string base_name() const;
+};
+
+struct OpenFlags {
+  // Whether to transparently compress or decompress, based on file extension
+  // or other heuristics.
+  bool transparent_compression = true;
+
+  OpenFlags() = default;
+};
+
+struct Selector {
+  // The folder in which to select files.
+  std::string base_folder;
+  // Whether to recurse into subfolders.
+  bool recursive = false;
+
+  MatchOptions() = default;
+};
+
+
+class FileSystem {
+ public:
+  virtual ~FileSystem();
+
+  /// Get statistics for the given target.
+  ///
+  /// Any symlink is automatically dereferenced, recursively
+  virtual Status GetTargetStats(const std::string& path, FileStats* out);
+  /// Same, for many targets at once.
+  virtual Status GetTargetStats(const std::vector<std::string>& paths,
+                                std::vector<FileStats>* out) = 0;
+  virtual Status GetTargetStats(const Selector& select,
+                                std::vector<FileStats>* out) = 0;
+
+  /// Create a folder and subfolders.
+  virtual Status CreateFolder(const std::string& path, bool recursive = true) = 0;
```

Comment:
what are you throughts on the implementation for this, when the underlying store doesn't support folders explicitly (e.g. S3)?

Reply:
I think we should perhaps emulate them using `/` as a standard delimiter (perhaps configurable when instantiating the S3 filesystem object).

Reply:
This conversation has been had many times over!

Specifically for S3, there are two ideas of what a folder might mean, in the absence of a real posix-like hierarchy: the simplest, that create-folder is a no-op, and you only ever infer folders if they contain things; and that a empty key which ends with '/' is to be considered an empty folder (but it could later morph to a file if data is written). The latter is the convention used by the S3 console.

Reply:
Well, I've never used S3, so I can't make a choice myself. The question would be: how do people usually use buckets?

Reply:
> how do people usually use buckets

You are asking a different question here! Buckets are the one level of "directory" that do exist, with some strange caveats (permissions work differently)

Note that in S3, it is possible to have a "directory" (key) and a "file" (prefix) with the same name, so isdir() and isfile() would both be True.

Reply:
Ok, so my question is: assuming someone wants to store some kind of filesystem hierarchy in a S3 bucket, how would they usually do it?

Reply:
Listing files with the S3 API allows for a delimiter/prefix mechanism, you say how you want to define the directory name separators (always "/") and what prefix you want to list, and you get back a list of keys with that prefix and no more delimiters and a list of "common prefixes" of keys with that prefix and more delimiters. That acts a lot like a posix list directory.

Reply:
Ok, so basically using `/` as a delimiter to simulate directories in S3 is an appropriate approach?

Reply:
>  using / as a delimiter to simulate directories in S3

Definitely, but it needs to be done with care, mindful that it isn't truly like that (and this is the case for other key-value stores too)

Reply:
Right, but that's more of a quality of implementation issue.

Reply:
Good testing, of course; but an outstanding issue on s3fs has been whether `mkdir(path)` followed by `exists(path)` should necessarily always return True. It is debatable!